Integrate Synthswap and 1inch swaps

avclarke opened the pull request:
## Description
Enable users to swap synth to erc20 via synthswap and other tokens via 1inch 

## Related issue
https://github.com/Kwenta/kwenta/issues/951

platschi commented:
On a first glance looks great! Will add more comments while testing.

1) Search bar only searches synths, not tokens list. We should probably rename it to "Search" and make sure it searches all lists.

<img width="616" alt="Screen Shot 2022-06-21 at 22 11 39" src="https://user-images.githubusercontent.com/548702/174889128-e9fcf469-7e1b-4368-a922-897f2e6fed31.png">

platschi commented:
From #951, missing (still opens Uniswap widget):

> "Get more sUSD" widget currently uses Uniswap widget; remove this and forward the button to our exchange once 1inch is integrated

platschi commented:
3) Synth icon is not pulled when trying to swap on vercel deploy. They are however visible for tokens, but not for other synths. See screenshot.

<img width="637" alt="Screen Shot 2022-06-21 at 23 01 19" src="https://user-images.githubusercontent.com/548702/174896722-71fbaf2d-7fed-40aa-8eab-e1d967bb9c2d.png">

4) "Exchange Fee" and "Fee Cost" are not visible when swapping from/to tokens. I guess if no (protocol) fees apply, we should just leave the label text and show a placeholder `-` for the values, to keep the UI consistent?

avclarke commented:
Thanks for the review.

1. Good catch will fix the search
2. ok
3.1. I'm not seeing the issue with synth icons on vercel deploy, seem to be showing up ok for me
3.2. Will look into gas fee issue
4. There will be some fees here but it needs some refactoring on the fee calculation in useExchange to support synthswap. I was thinking we could do that in a separate task. For now I'll make it consistent.

avclarke commented:
@platschi I do see the issue with icon sometimes actually, will also look into that

avclarke commented:
@platschi I've addressed all comments apart from the fees. I've left it as is, to hide fees on non synth to synth swap as this was the existing logic. As mentioned previously we'll need to refactor a bit to populate fees section for synthswap, I'm happy to add to this PR but as I will be off from tomorrow to Tuesday I don't think it will be a blocker for merging this feature if we wanted to break fees into a separate task.

platschi commented:
Sometimes I see this yellow `-100%` indicator but not sure what's that. Seems like the swap values are fine

<img width="637" alt="Screen Shot 2022-06-22 at 17 02 30" src="https://user-images.githubusercontent.com/548702/175063689-4db174f1-2f7b-4acd-87b7-006e56a9e178.png">

avclarke commented:
> Sometimes I see this yellow `-100%` indicator but not sure what's that. Seems like the swap values are fine
> 
> <img alt="Screen Shot 2022-06-22 at 17 02 30" width="637" src="https://user-images.githubusercontent.com/548702/175063689-4db174f1-2f7b-4acd-87b7-006e56a9e178.png">

@platschi this is the slippage percent, and it was showing -100% because coingecko price fetching was hardcoded to ethereum and failing to fetch prices of optimism addresses. Should be fixed now as updated the query to include the correct network.

platschi commented:
- When selecting e.g. sUSD -> AELIN, I cannot type in values in the "Into" input field.  However, this works (as expected) if I switch to sUSD -> sLINK, for example. Would expect to be able to input values in both fields.

- sUSD -> ELK does not calculate a value for ELK
- sUSD -> WBTC, sUSD -> USDT, sUSD -> USDC values are extremely low  and/or with a 100% slippage, probably no liquidity here, do we still want to offer these tokens or hide them from the list given it's basically useless to swap them?

avclarke commented:
>When selecting e.g. sUSD -> AELIN, I cannot type in values in the "Into" input field. However, this works (as expected) if I switch to sUSD -> sLINK, for example. Would expect to be able to input values in both fields.

I forgot to mention I disabled input on "into" field for synthswap assets as it's going to be a bit more effort to make the conversion the other way. Could we break into another ticket?

> sUSD -> ELK does not calculate a value for ELK
I just checked and this gives an insufficient liquidity error. Do we have a standard error notification or something we can show for things like this?

>sUSD -> WBTC, sUSD -> USDT, sUSD -> USDC values are extremely low and/or with a 100% slippage, probably no liquidity here, do we still want to offer these tokens or hide them from the list given it's basically useless to swap them?
This is a good catch and actually because I'm not using the correct asset decimals

>Showing Slippage is cool, but we should maybe make it clear that the yellow % value refers to Slippage. It wasn't immediately clear to me when I saw it first :) Maybe add a "Slippage: " string in front of them, or something similar that fits
Agree this is not very obvious, it was just in the existing UI. I will see how it looks with "slippage" label.

platschi commented:
> > When selecting e.g. sUSD -> AELIN, I cannot type in values in the "Into" input field. However, this works (as expected) if I switch to sUSD -> sLINK, for example. Would expect to be able to input values in both fields.
> 
> I forgot to mention I disabled input on "into" field for synthswap assets as it's going to be a bit more effort to make the conversion the other way. Could we break into another ticket?

All right will create another ticket to fix it.

> > sUSD -> ELK does not calculate a value for ELK
> 
> I just checked and this gives an insufficient liquidity error. Do we have a standard error notification or something we can show for things like this?

Currently just showing the error warning within the "Swap" button, you can test by forcing an insufficient balance error. We can place it there for now until we have a better error handling in place (part of #1052, #491)

> 
> > sUSD -> WBTC, sUSD -> USDT, sUSD -> USDC values are extremely low and/or with a 100% slippage, probably no liquidity here, do we still want to offer these tokens or hide them from the list given it's basically useless to swap them?
> 
> This is a good catch and actually because I'm not using the correct asset decimals

:+1:

> > Showing Slippage is cool, but we should maybe make it clear that the yellow % value refers to Slippage. It wasn't immediately clear to me when I saw it first :) Maybe add a "Slippage: " string in front of them, or something similar that fits
> 
> Agree this is not very obvious, it was just in the existing UI. I will see how it looks with "slippage" label.

Thanks!

Tburm commented on containers/Convert/Convert.tsx:
```diff
@@ -3,16 +3,19 @@ import axios from 'axios';
-
-import erc20Abi from 'lib/abis/ERC20.json';
+//@ts-ignore
+import getFormatedSwapData from '@jaredborders/synthswap-utils';
```
Should we find a way to publish this from Kwenta? @JaredBorders

JaredBorders replied:
We can. Will put on todo list for tmr

avclarke replied:
Also could be good to add the types or convert the code to ts, maybe a nice devdao task

JaredBorders replied:
Latest version of Synthswap package no longer uses that

JaredBorders replied:
1.0.3

Tburm commented on queries/coingecko/useCoinGeckoTokenPricesQuery.ts:
```diff
@@ -1,24 +1,29 @@
 				)}&vs_currencies=usd`
 			);
-
+			console.log('response', response);
```
Can remove

Tburm commented on sections/shared/modals/SelectCurrencyModal/SelectCurrencyModal.tsx:
```diff
@@ -106,88 +111,196 @@ export const SelectCurrencyModal: FC<SelectCurrencyModalProps> = ({
+									{assetSearch ? (
+										<span>{t('modals.select-currency.header.search-results')}</span>
+									) : synthCategory != null ? (
+										t('modals.select-currency.header.category-synths', {
```
This will display "Crypto Synths" or "Forex Synths" above the 1inch tokens when there is a filter selected

<img width="422" alt="image" src="https://user-images.githubusercontent.com/10401554/176785531-26960898-100e-4658-81f0-255191af2c8c.png">

avclarke replied:
Good spot, I think I'll hide 1inch tokens when forex is selected and then update the label for when crypto is selected

Tburm reviewed (commented):
Swapping from ETH -> sUSD gave me an error. Maybe we need to handle the ETH case differently than other ERC20s?

<img width="322" alt="image" src="https://user-images.githubusercontent.com/10401554/176785793-c3f86c68-1a7d-4d12-bfaa-723f18687d20.png">

Tburm commented:
This title also looks odd for non-sUSD pairs...

<img width="238" alt="image" src="https://user-images.githubusercontent.com/10401554/176786620-c1b5ab80-d7fd-4250-a3c7-cbd130ed4661.png">

For reference, an sUSD pair:
<img width="237" alt="image" src="https://user-images.githubusercontent.com/10401554/176786656-884f2db2-c651-4a58-8e17-ad685718326b.png">

avclarke commented:
> This title also looks odd for non-sUSD pairs...
> 
> <img alt="image" width="238" src="https://user-images.githubusercontent.com/10401554/176786620-c1b5ab80-d7fd-4250-a3c7-cbd130ed4661.png">
> 
> For reference, an sUSD pair: <img alt="image" width="237" src="https://user-images.githubusercontent.com/10401554/176786656-884f2db2-c651-4a58-8e17-ad685718326b.png">

I was just looking into this and it seems an existing issue in `formatCryptoCurrency`, it's adding currency key as prefix and sign as suffix. I will switch them unless there is a reason for this that I am missing.

<img width="687" alt="Screenshot 2022-07-01 at 12 32 48" src="https://user-images.githubusercontent.com/3802342/176887265-c92fae20-e208-4976-b685-50e1306c7373.png">
